Add the ability to register additional controllers

hatched opened the pull request:
## Done

Adds the ability to register additional controllers on the controllers page and have that controllers models added to the model list page.

## QA

Bootstrap two separate controllers on your provider of choice...
- `juju bootstrap aws test1`  
- `juju bootstrap aws test2`  


### JAAS
- Setup the dashboard for JAAS using [this guide](https://github.com/canonical-web-and-design/jaas-dashboard#in-jaas)
- Ensure that everything works as expected for the usual JAAS experience.
- Add your controllers
  - Switch to the Controllers page and click the "Register a Controller" button.
  - Follow the steps in the application.
  - Refresh
- Your `default` and `controller` models for both controllers should now be in the models list
- The two controllers should now be shown in the "Registered Controllers" section.

### Juju
- Build the tarball and update your `test1` controller from this branch with [this guide](https://github.com/canonical-web-and-design/jaas-dashboard#uploading-the-tarball-to-a-juju-controller)
- Add the `test2` controller and the `jaas` controller using the steps above.
- Your `default`, `controller`, and JAAS models for both controllers should now be in the models list.
- The two controllers should now be shown in the "Registered Controllers" section.

## Details

Still to do:
- [x] Update tests

For a follow-up:
- [ ] Do not require refresh after adding a controller
- [x] UI Enhancements.
- [ ] Removal of controllers from the list.
- [ ] Display controller-level errors in the controller page.
- [ ] Address remaining XXX comments.
- [ ] Show all registered controllers, even if not connected.
- [ ] JAAS will be adding a 'default' controller so that one shows up in the list. For now this will be blank until JAAS is updated with this feature.
- [x] Fix counts in charts.
- [ ] Validate form input for controller registration form.
- [ ] Add the `isJuju` check on a controller level basis.
- [ ] Ask users if they would like to clear saved controllers on log-out as the credentials are stored locally.
- [ ] How do we indicate to the user that the localstorage data for controllers is bad and needs to be cleared? [comment](https://github.com/canonical-web-and-design/jaas-dashboard/pull/594/files#r457521758)
- [ ] Use the submitted controller name instead of the controller `path` in the UI.
- [ ] Implement the [accept cert flow](https://github.com/canonical-web-and-design/jaas-dashboard/pull/594#issuecomment-662511683)

Fixes #554
Fixes #589 
Fixes #565 

## Screenshots

<img width="1680" alt="Screen Shot 2020-07-16 at 2 47 28 PM" src="https://user-images.githubusercontent.com/532033/87721221-546fc900-c773-11ea-99b1-ebd1800f7888.png">

<img width="562" alt="Screen Shot 2020-07-22 at 12 39 45 PM" src="https://user-images.githubusercontent.com/532033/88215205-89b76380-cc18-11ea-9b29-50ff40a3fe81.png">

webteam-app commented:
Starting demo at: https://jaas-dashboard-canonical-web-and-design-pr-594.run.demo.haus/

hatched commented on src/app/check-auth.js:
```diff
@@ -23,27 +24,56 @@ const actionWhitelist = [
+      wsControllerURL: The full controller websocket url that the controller
+        is stored under in redux in order to determine it's logged in status.
+*/
+export default ({ getState }) => (next) => async (action, options) => {
```
This change adds an additional argument to the `dispatch` method. It's call signature now requires the `wsControllerURL` to perform actions that involve a specific controller.
```js
dispatch(action, options)
```

hatched commented on src/components/SlidePanel/SlidePanel.js:
```diff
@@ -0,0 +1,22 @@
+
+import "./_slide-panel.scss";
+
+function SlidePanel({ children, onClose }) {
```
This is just a basic pass at this panel, we'll want to enhance its functionality in follow-ups.

hatched commented on src/components/SlidePanel/_slide-panel.scss:
```diff
@@ -0,0 +1,15 @@
+  height: calc(100% - 25px);
+  position: absolute;
+  right: 0;
+  top: 24px; // height of the top bar
```
This is flaky because every time we update the height of the top bar this will have to be changed... maybe there is a better approach to this so that it works dynamically across the application?

hatched commented on src/index.js:
```diff
@@ -40,16 +38,19 @@ if (process.env.NODE_ENV === "production") {
-    setTimeout(checkConfigExists, 500);
-    return;
+  if (!window.jujuDashboardConfig) {
+    if (checkCounter < 5) {
```
This is a driveby fix to fix #589

barrymcgee replied:
Can we cherrypick this out into a separate branch and get it landed?

barrymcgee commented on src/app/model-poller.js:
```diff
@@ -21,46 +21,72 @@ import {
-      });
-    } while (isLoggedIn(reduxStore.getState()));
+
+      if (process.env.NODE_ENV === "production") {
```
We do this check before initing Sentry at root level.. is it also needed here? If so, what can we do at the conditional at root level to ensure calling `Sentry` just gracefully fails if not on production?

hatched replied:
We need to connect to the controller first to be able to get its version so it has to be done here.

We could probably create a wrapper around sentry which would have this check built in. But that would be for a follow-up, feel free to file an issue for that.

barrymcgee commented on src/app/model-poller.js:
```diff
@@ -21,46 +21,72 @@ import {
+        );
+        // XXX the isJuju Check needs to be done on a per-controller basis
+        if (!isJuju) {
+          // This call will be a noop if the user isn't an administrator
```
a noop?

hatched replied:
It means "no operation". Something is a noop if it doesn't do anything.

spencerbygraves commented:
@hatched thanks, a couple of things...

- The header bar is too high, which is the same on the live site. Should that be a separate issue?
 
![image](https://user-images.githubusercontent.com/9105999/87683209-b2eb7600-c778-11ea-947f-be6d20cb0f60.png)

- We have two "model status across controllers" titles, can you remove the second one please
- @clagom should the register button be below the table header?
- The button should be sentence case, no upper case "C"
- Did you use `p-icon--help` from [here](https://vanillaframework.io/docs/patterns/icons)? The colour is wrong

Thanks

clagom commented:
I think the button could be above the header, yes. To be honest I'm still not 100% convinced of its position, but it's a start. @spencerbygraves @hatched

hatched commented:
@spencerbygraves I've already planned for there to be a follow-up to address the UI issues as this branch is sufficiently large enough.
 
>     * The header bar is too high, which is the same on the live site. Should that be a separate issue?

This is how it is in the master branch. If you'd like to change it feel free to file an issue and we can address it separately.

>     * We have two "model status across controllers" titles, can you remove the second one please

Oops, yep consider it gone

>     * @clagom should the register button be below the table header?

I'll leave the button where it is and we can move it once the design is ironed out.  

>     * The button should be sentence case, no upper case "C"

Done.

>     * Did you use `p-icon--help` from [here](https://vanillaframework.io/docs/patterns/icons)? The colour is wrong

I did, but it's way to dark (see email thread) So I chose the lighter of the two colour options.

hatched commented:
@spencerbygraves  I went ahead and updated some of the UI elements here....any further changes can wait for a follow-up.

>     * The header bar is too high, which is the same on the live site. Should that be a separate issue?

Fixed

>     * @clagom should the register button be below the table header?

I moved it into the header as in one of the designs, it looked better there.

I also updated the charts, see the screenshot in the top post.

spencerbygraves commented:
>I did, but it's way to dark (see email thread) So I chose the lighter of the two colour options.

I didn't see an email thread about icon colours. There's two variants, light and dark theme. It should look like this:

![image](https://user-images.githubusercontent.com/9105999/87790623-c2ca8f00-c838-11ea-8235-2f13edb5fe33.png)

anthonydillon commented on src/app/actions.js:
```diff
@@ -139,6 +160,19 @@ export function logOut(getState) {
+      }
+    } catch (e) {
+      // XXX Add to Sentry.
+      console.log("oops bad data in the additionalControllers localStorage");
```
Might not just be bad data, maybe something like "Error retrieving additional controllers". Then is it worth clearing the localstorage to regain a working UI without the user having the clear the storage themselves?

hatched replied:
This is an interesting problem...If there is bad data they would have to go in and manually edit/clear it out which isn't a great UX. However I don't want to just clear it as then all evidence of what caused the issue would be erased. 

I'll update the error message here and add another todo around surfacing this to the user somehow so that they can take appropriate action.

anthonydillon commented on src/components/SlidePanel/SlidePanel.js:
```diff
@@ -0,0 +1,22 @@
+function SlidePanel({ children, onClose }) {
+  return (
+    <div className="slide-panel">
+      <div className="slide-panel--close">
```
`slide-panel__close` this is an element of the slide-panel not a modifier to it.

hatched replied:
removed this in favour of your other comment.

anthonydillon commented on src/components/SlidePanel/SlidePanel.js:
```diff
@@ -0,0 +1,22 @@
+          tabIndex="0"
+        ></span>
+      </div>
+      <div className="slide-panel--content">{children}</div>
```
`slide-panel__content`

hatched replied:
done

anthonydillon commented on src/components/SlidePanel/_slide-panel.scss:
```diff
@@ -0,0 +1,15 @@
+.slide-panel {
+  background-color: white;
```
background-color: $color-x-light;

hatched replied:
done

anthonydillon commented on src/components/SlidePanel/_slide-panel.scss:
```diff
@@ -0,0 +1,15 @@
+.slide-panel {
+  background-color: white;
+  box-shadow: 0 1px 5px 1px rgba(51, 51, 51, 0.2);
```
`box-shadow: $box-shadow`

hatched replied:
done

anthonydillon commented on src/components/SlidePanel/_slide-panel.scss:
```diff
@@ -0,0 +1,15 @@
+  height: calc(100% - 25px);
+  position: absolute;
+  right: 0;
+  top: 40px; // height of the top bar
```
If your going to start 40px down you need to minus that from the height. Therefore the height value above should be `calc(100% - 40px)`. As they are the same you could use a variable to keep them aligned and easier read. Removing the need for the comment.

hatched replied:
Done

anthonydillon commented on src/components/SlidePanel/SlidePanel.js:
```diff
@@ -0,0 +1,22 @@
+function SlidePanel({ children, onClose }) {
+  return (
+    <div className="slide-panel">
+      <div className="slide-panel--close">
```
Please use a button similar to the model pattern:
```html
<button class="p-modal__close" aria-label="Close active modal" aria-controls="modal">Close</button>
```
Then you want need to worry about `role`, `tabIndex` or the cursor styling.

hatched replied:
Good call, done

anthonydillon commented on src/juju/actions.js:
```diff
@@ -109,9 +110,9 @@ export function fetchModelStatus(modelUUID) {
-    const controllers = getState()?.juju?.controllers;
+    const controllers = getState()?.juju?.controllers[wsControllerURL];
     const model = modelInfo.results[0].result;
     const updatedControllers = cloneDeep(controllers).map((controller) => {
```
You might want to check controllers is not null now your referencing an item in the array

hatched replied:
Good call, done.

anthonydillon commented on src/pages/Controllers/Controllers.js:
```diff
@@ -72,54 +77,320 @@ function Details() {
   return (
     <>
+      <div className="controllers--header">
+        <div className="controllers--heading">
```
These should be `__` to be BEM

hatched replied:
done

anthonydillon commented on src/pages/Controllers/_controllers.scss:
```diff
@@ -1,9 +1,63 @@
 .controllers {
+  th {
+    padding-bottom: 6px;
+    padding-top: 6px;
```
Why?

hatched replied:
The header styles by default have too much padding on the top so this makes the spacing consistent between the charts and two tables.

hatched commented:
Thanks for the reviews @anthonydillon @barrymcgee The requested changes have been made and it's ready for another look.

spencerbygraves commented:
@hatched thanks, can we change a few things please, I've added a design below.

- Align the title and form labels
- Align close icon to fields
- Move green buttons
- Add hr above green buttons
- The check box text should be normal not small

![image](https://user-images.githubusercontent.com/9105999/88081606-87053180-cb78-11ea-81cb-bfd1b278b50e.png)

anthonydillon commented:
QA points captured in a Meet.

spencerbygraves commented:
@hatched I've updated the design:

![image](https://user-images.githubusercontent.com/9105999/88083852-73a79580-cb7b-11ea-9667-dc43cdba4768.png)

clagom commented:
> @hatched I've updated the design:
> 
> ![image](https://user-images.githubusercontent.com/9105999/88083944-8f12a080-cb7b-11ea-9603-3f51e59103f7.png)

Are these 2 steps? Not a blocker for v1, but it just makes me wonder what I need to do as a user :D

hatched commented:
Thanks for the designs @spencerbygraves @clagom, I took a few liberties and put the updated image in the description of this PR.

spencerbygraves commented:
@hatched @clagom I don't mind combining the title and text for the checkboxes, but I have one question.

Originally we had the text "Visit the controller to accept..." before the checkbox, now it's the other way around. Is it dependant on accepting the certificate first?

clagom commented:
> @hatched @clagom I don't mind combining the title and text for the checkboxes, but I have one question.
> 
> Originally we had the text "Visit the controller to accept..." before the checkbox, now it's the other way around. Is it dependant on accepting the certificate first?

Yes that's also what I was thinking. It's unclear to me what the order is, should I go and accept the certificate and then tick the checkbox? The wording doesn't probably point out **first** that you need to go and accept before registering.

spencerbygraves commented:
@clagom @hatched could this work? I also updated the wording slightly.

![image](https://user-images.githubusercontent.com/9105999/88160304-5ff54100-cc06-11ea-9577-4d308b6a9b60.png)

clagom commented:
Thanks @spencerbygraves I think this is clearer. What do you think about adding an 'attention' kind of icon? too much?
<img width="589" alt="Screenshot 2020-07-22 at 10 56 12" src="https://user-images.githubusercontent.com/1953388/88163022-fbd47c00-cc09-11ea-9d6a-a1f85cfe5d8a.png">

spencerbygraves commented:
@clagom looks good

anthonydillon commented:
I think we should show the address so the user knows where they are going. We can also do something smart with `:visited`

anthonydillon commented:
Also, can we disable the button until we have a validate form?/

clagom commented:
Both good points

anthonydillon commented:
Not for this round by it might be nice to show this as a authentication type of UI.

For example:
![IMG_20200722_160844](https://user-images.githubusercontent.com/1413534/88193936-20dee400-cc36-11ea-8354-84bb810996f4.jpg)

hatched commented:
All requests have been addressed, it's ready for another review/qa

anthonydillon commented on src/pages/Controllers/_controllers.scss:
```diff
@@ -1,9 +1,93 @@
+  .controllers--registered-tooltip {
+    @include vf-icon-help($color-mid-light);
+
+    cursor: pointer;
```
Can you change this it `help` I clicked it and expected it to do something.

hatched replied:
Good call, done

anthonydillon commented on src/pages/Controllers/_controllers.scss:
```diff
@@ -1,9 +1,93 @@
+  }
+
+  .slide-panel__content {
+    width: 550px;
```
Can you change this to `max-width` (for small screens)

hatched replied:
done

anthonydillon commented on src/components/SlidePanel/_slide-panel.scss:
```diff
@@ -0,0 +1,21 @@
+  overflow: scroll;
+  position: fixed;
+  right: 0;
+  top: $top-header-height;
```
Can you add margin-left that matches the width of the collapsed navigation panel so this UI works nicely on small screen.

hatched replied:
done

hatched commented:
Thanks everyone for all the feedback on this PR!

:shipit:
